src/hardware/mixer.cpp

```diff
@@ -406,160 +404,150 @@ void MixerChannel::AddSamples(uint16_t len, const Type *data)
 		// prevent severe clicks and pops. Becomes a no-op when done.
-		envelope.Process(stereo, interpolate, prev_sample, next_sample);
+		envelope.Process(stereo, prev_sample);
 
-		//Where to write
-		mixpos &= MIXER_BUFMASK;
-		if (!interpolate) {
+		auto left = prev_sample[mapped_channel_left] * volmul[0];
+		auto right = (stereo ? prev_sample[mapped_channel_right]
+		                     : prev_sample[mapped_channel_left]) *
+		             volmul[1];
 
-			mixer.work[mixpos][mapped_output_left] +=
-			        prev_sample[mapped_channel_left] * volmul[0];
+		out_frame = {0.0, 0.0};
+		out_frame[mapped_output_left] += left;
+		out_frame[mapped_output_right] += right;
 
-			mixer.work[mixpos][mapped_output_right] +=
-			        (stereo ? prev_sample[mapped_channel_right]
-			                : prev_sample[mapped_channel_left]) *
-			        volmul[1];
+		out.push_back(out_frame[0]);
+		out.push_back(out_frame[1]);
 
-		} else {
-			const auto diff_mul = freq_counter & FREQ_MASK;
-			auto sample = prev_sample[mapped_channel_left] +
-			              (((next_sample[mapped_channel_left] - prev_sample[mapped_channel_left]) *
-			                diff_mul) >>
-			               FREQ_SHIFT);
-			mixer.work[mixpos][mapped_output_left] += sample * volmul[0];
-			if (stereo) {
-				sample = prev_sample[mapped_channel_right] +
-				         (((next_sample[mapped_channel_right] - prev_sample[mapped_channel_right]) *
-				           diff_mul) >>
-				          FREQ_SHIFT);
-			}
-			mixer.work[mixpos][mapped_output_right] += sample * volmul[1];
-		}
-		//Prepare for next sample
-		freq_counter += freq_add;
-		mixpos++;
-		done++;
+		++pos;
 	}
-	
+}
+
+template <class Type, bool stereo, bool signeddata, bool nativeorder>
+void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
+{
+	assert(frames <= MIXER_BUFSIZE);
+	MIXER_LockAudioDevice();
+	last_samples_were_stereo = stereo;
+
+	auto &convert_out = resample ? mixer.resample_temp : mixer.resample_out;
+	ConvertSamples<Type, stereo, signeddata, nativeorder>(data, frames, convert_out);
+
+	spx_uint32_t out_frames;
+	if (resample) {
+		spx_uint32_t in_frames = mixer.resample_temp.size() / 2;
+		out_frames = mixer.resample_out.capacity() / 2;
+		speex_resampler_process_interleaved_float(resampler,
+		                                          mixer.resample_temp.data(),
+		                                          &in_frames,
+		                                          mixer.resample_out.data(),
+		                                          &out_frames);
+	} else {
+		out_frames = mixer.resample_out.size() / 2;
+	}
+
+	// Mix channel to the master output
+	auto mixpos = check_cast<work_index_t>(mixer.pos + done);
+
+	work_index_t pos = 0;
+	auto frames_left = out_frames;
+	do {
+		mixpos &= MIXER_BUFMASK;
+		mixer.work[mixpos][0] += mixer.resample_out[pos++];
+		mixer.work[mixpos][1] += mixer.resample_out[pos++];
+		++mixpos;
+	} while (--frames_left);
+
+	done += out_frames;
```

Comment:
Well, I have a nicer solution to this, but it doesn't _quite_ work (I can hear the audio at the correct pitch and speed, but there's some crackling as well). I'm fairly new to STL, so maybe a C++ guru among you can point out my mistake, I just couldn't figure out what's wrong with it.

```cpp
if (resample) {
    spx_uint32_t in_frames = mixer.resample_temp.size() / 2;
    spx_uint32_t out_frames = mixer.resample_out.capacity() / 2;
    speex_resampler_process_interleaved_float(resampler,
                                              mixer.resample_temp.data(),
                                              &in_frames,
                                              mixer.resample_out.data(),
                                              &out_frames);
    auto newsize = out_frames * 2;
    mixer.resample_out.resize(newsize);
}

// Mix channel to the master output
auto pos = mixer.resample_out.begin();
auto mixpos = check_cast<work_index_t>(mixer.pos + done);

while (pos != mixer.resample_out.end()) {
    mixpos &= MIXER_BUFMASK;
    mixer.work[mixpos][0] += *pos++;
    mixer.work[mixpos][1] += *pos++;
    ++mixpos;
}

auto out_frames = mixer.resample_out.size() / 2;
done += out_frames;
```

Reply:
In the nicer solution, use `.size()` instead of `.capacity()`.
Although `Capacity` _does_ tell you how much allocated memory exists, it's implementation specific and merely has to be something as big or more than the logical size (perhaps some systems could give you 2^32 all the time). To be more specific, `Capacity()` can exceed `::end() - ::begin()`, which are the logical limits.

Always use Size() and Resize(..) because they manage the logical extent, and match what `.begin()` and `.end()` will give you. 

Here's an example I threw together:

``` c++
#include <stdio.h>
#include <vector>

void print_data(const std::vector<int> & buf) {
    printf("\nsize() = %lu\n", buf.size());
    printf("capacity() = %lu\n", buf.capacity());
    printf("data() = {");
    const auto data = buf.data();
    for (auto i = 0; i < buf.capacity(); ++i)
        printf("%d ", data[i]);
    printf("}\n");
}

int main()
{
    std::vector<int> buf(10);
    for (auto i = 0; i < buf.size(); ++i)
        buf[i] = i;

    print_data(buf);

    buf.resize(5);
    print_data(buf);

    buf.resize(10);
    print_data(buf);

    return 0;
}
```

``` text
size() = 10
capacity() = 10
data() = {0 1 2 3 4 5 6 7 8 9 }

size() = 5
capacity() = 10
data() = {0 1 2 3 4 5 6 7 8 9 }

size() = 10
capacity() = 10
data() = {0 1 2 3 4 0 0 0 0 0 }
```

https://onlinegdb.com/KZujUccQC

Notice the resize() will zero-out the values between the previous smaller size up to the newer larger size. 

So I think that's what's happening here: we're telling the resampler it's free to write out `capacity() / 2` frames into the array. Then we resize() to match however many samples it wrote.  If the size() going in was less than capacity(), then the delta samples are going to be zeroed (like in the example).

Reply:
In the nicer implementation, if we take the non-resampling path (which I've edited-out):

``` c++
void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
{
	assert(frames <= MIXER_BUFSIZE);
	MIXER_LockAudioDevice();
	last_samples_were_stereo = stereo;

	auto &convert_out = mixer.resample_out;
	ConvertSamples<Type, stereo, signeddata, nativeorder>(data, frames, convert_out);

	auto pos = mixer.resample_out.begin();
	auto mixpos = check_cast<work_index_t>(mixer.pos + done);

	while (pos != mixer.resample_out.end()) {
		mixpos &= MIXER_BUFMASK;
		mixer.work[mixpos][0] += *pos++;
		mixer.work[mixpos][1] += *pos++;
		++mixpos;
	}

	auto out_frames = mixer.resample_out.size() / 2;
	done += out_frames;
```

If the incoming number of `frames` is less than the size of `resample_out` vector, then `ConvertSamples()` will only populate `frames`-worth of samples into ` the `resample_out` vector.

So then when we get to the `while` loop, we need to be sure only `frames`-worth of samples are written into the mixer's buffer instead of the entire `resample_out` vector (currently the nicer implementation will crawl from `.begin()` to `.end()` of that vector.

Reply:
> In the nicer solution, use `.size()` instead of `.capacity()`

I could do that, but it's not ideal. I'm only using `vector` instead of `array` because the amount of data the resampler outputs is variable (dependent on the in and out rates), and I wanted to bundle `out_size` with `out_data`, so to speak. I've read up on it, and if you provide the size of the vector at construction time, and the vector never grows in size, effectively you'll get an array that "knows" it's actual size, but reallocations will never happen (as long as you don't resize it above it's initial capacity).

So when passing data to the resampler, I need `capacity` to inform it that it can use the whole buffer, then on the way out I need to set the actual size with `resize`. After that, `size` would return the actual size. Well, that's how it's supposed to work at least, based on my understanding.

> Although Capacity does tell you how much allocated memory exists, it's implementation specific and merely has to be something as big or more than the logical size (perhaps some systems could give you 2^32 all the time).

That's [`max_size`](https://www.cplusplus.com/reference/vector/vector/max_size/) what you're referring to, I think.

[`capacity`](https://www.cplusplus.com/reference/vector/vector/capacity/) should work as I described above.

> Capacity() can exceed ::end() - ::begin(), which are the logical limits.

Yeah that's true, but in cases like this, we need it so the C code can fill up the whole buffer, potentially, then we set the real size with `resize`.

Anyway, I'm starting to doubt whether this whole approach is worth it, and maybe I should just simply use `array`. I thought it's nicer not to have separate "data" and "size" vars, but we're not dealing with tens or hundreds of buffers here so it wouldn't be that much more confusing or error prone.

Reply:
> in cases like this, we need (`capacity`) so the C code can fill up the whole buffer, potentially, then we set the logical size with `resize`.

The problem is that `resize` will zero the range of values when going from `size` up to the resized (potentially full `capacity`).

So that's how size impacts us.

Let's say the vector is created with a large initial size, say 1024 samples. 

Assuming the mixer's running at a 48 KHz, a typical 1ms pass through will ask for 48 frames output from the resampler. 

Depending on jitter in DOSBox's own timing, sometimes it will want more or less than those 48 frames. 

Say our channel is running at 22 KHz, so for a 1ms pass we're typically resampling 22 frames up to 48 frames. Capacity is always 1024 samples. 

Let's say the current pass asks for 20 frames, which we resample up to say 42 frames. We resize to 84 samples. 

Next pass through, let's say we ask for 24 frames which we resample up to 50 frames. The out buffer's logical size is 84 samples, set from the prior pass (capacity is still 1024). The resampler just wrote 50 frames or 100 samples. Now we resize to 100 samples.  

At this point, the resize will zero the range from the 84th to 100th sample.

Reply:
> The problem is that resize will zero the range of values when going from size up to the resized (potentially full capacity).

Ah, this sentence made everything clear! I wasn't aware of the zeroing behaviour when resizing, and when you mentioned it before I didn't realise that the zeroing can also happen when _upsizing_, not just when downsizing. Rookie mistake, indeed!

The rest of what you wrote makes perfect sense.

Actually, the solution is quite simple: resize to capacity before calling the spex resampler, then downsizing will not zero any actual data. I don't the extra initialisation matters at all with such small buffers.

Reply:
I realized these resampling buffers are members of the `mixer`, so they're shared across the channels (which is fine). 

For channels running at the mixer's rate, without resampling, the output buffer will be sized to the number of output frames-per-millisecond, typically 48 frames.

For channels needing resampling, the  the output buffer will be resized up to `capacity`, which means zero'ing out 16384 * 2 channels worth of values (even though we also only need to write 48 frames). Given these are 32-bit floats, that means zero'ing 128 KB of memory per tick. Multiply by 1000 ticks/s brings up to 128 MB/s -- per channel (versus what's actually needed, which is writing ~48 frames per tick or  ~384 KB/s of resampled data).

We can use the DSP's `speex_resampler_get_ratio` to help us determine the maximum possible output size, given an input quantity.  So this way, our resize call is exactly as big as it needs. 

Here's an attempt to improve it (feel free to slice and dice it!):

``` diff
+spx_uint32_t estimate_max_out_frames(SpeexResamplerState *resampler,
+                                     const spx_uint32_t in_frames)
+{
+       assert(resampler);
+       assert(in_frames);
+
+       spx_uint32_t in = 0;
+       spx_uint32_t out = 0;
+       speex_resampler_get_ratio(resampler, &in, &out);
+
+       assert(in && out);
+       return ceil_udivide(in_frames * out, in);
+}
+
 template <class Type, bool stereo, bool signeddata, bool nativeorder>
 void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
 {
@@ -520,14 +534,12 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
 
        auto &convert_out = resample ? mixer.resample_temp : mixer.resample_out;
        ConvertSamples<Type, stereo, signeddata, nativeorder>(data, frames, convert_out);
+       assert(frames == convert_out.size() / 2); // 1:1 conversion so should match
 
        if (resample) {
                spx_uint32_t in_frames = mixer.resample_temp.size() / 2;
-               spx_uint32_t out_frames = mixer.resample_out.capacity() / 2;
-
-               // Make sure useful data won't get zeroed out when we resize the vector
-               // to the actual output length after resampling
-               mixer.resample_out.resize(mixer.resample_out.capacity());
+               spx_uint32_t out_frames = estimate_max_out_frames(resampler, in_frames);
+               mixer.resample_out.resize(out_frames * 2);
 
                speex_resampler_process_interleaved_float(resampler,
                                                          mixer.resample_temp.data(),
@@ -535,9 +547,10 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
                                                          mixer.resample_out.data(),
                                                          &out_frames);
 
-               // out_frames now contains the actual number of frames output by the resampler
-               auto newsize = out_frames * 2;
-               mixer.resample_out.resize(newsize);
+               // out_frames now contains the actual number of resampler frames,
+               // ensure the number of output frames is within the logical size.
+               assert(out_frames <= mixer.resample_out.size() / 2);
+               mixer.resample_out.resize(out_frames * 2); // only shrinks
        }
 
        // Mix channel to the master output
```

Reply:
OK, you're right, it's actually not insignificant when you put it that way. I'll do the suggested improvement.

Reply:
Actually, what you wrote made me think about the buffer size: why do we even need a 16k frame long buffer in the first place? Does that actually happen in practice that a pass can ask for that many frames? Seems unlikely to me.

What about the prebuffer setting? If I set that to say 30ms, that would introduce a 30ms delay to the sound, right? I guess to be able to set 100-150ms predelays, we'd need the large mixbuffer, after all.

Reply:
Indeed, the channel "millisecond buffer"  can be really small; in the GUS for example, the buffer we pass to `AddSamples` is at-most 48 frames, and if the callback asks for more then we keep looping until done:

``` c++
void Gus::AudioCallback(const uint16_t requested_frames)
{
	uint16_t generated_frames = 0;
	while (generated_frames < requested_frames) {
		...
		channel->AddSamples_s16(frames, play_buffer.data());
		CheckVoiceIrq();
		generated_frames += frames;
	}
}
```

The same chunked approach could be done in `AddSamples`:

``` c++
void MixerChannel::AddSamples(const uint16_t incoming_frames, const Type *data)
{
	uint16_t mixed_frames = 0;
	while (mixed_frames < incoming_frames) {
		... // digest in chunks
}
```

The `data`-pointer would need to be walked ahead: ConvertSamples<Type, stereo, signeddata, nativeorder>(**data**, frames, convert_out);, which I think is the tricky part, because that source data might be stereo, or not, and the type sizes can changes (before the conversion).